### PR TITLE
rewrite in typescript

### PR DESCRIPTION
**Pull request by ryantate13**
Fix for issue 1.

https://github.com/hustcc/variable-type/issues/1

It's a breaking change for non-Typescript users. Have to switch up syntax to reflect the default export so I thought a bump to 1.0 was appropriate.

```js
// from
const VT = require('variable-type');
// to
const VT = require('variable-type').default
```

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,16 +1,7 @@
...
-  - '2'
-  - '3'
-  - '4'
-  - '4'
-  - '5'
-  - '6'
-  - '7'
-  - '8'
+  - '10'
+  - '12'
```
Comment (hustcc) [inline .travis.yml:5]: keep nodejs 8 ~ 12.

File: README.md [hunk 2]
```diff
@@ -15,9 +15,9 @@
 
 Then import it.
 
-```js
-import VT from 'variable-type'; // ES6
-var VT = require('variable-type'); // ES5 with npm
+```typescript
+import VT from 'variable-type'; // ES6/Typescript
+const VT = require('variable-type').default; // ES5 with npm
```
Comment (hustcc) [inline README.md]: keep import, remove require.

File: README.md [hunk 3]
```diff
@@ -15,9 +15,9 @@
 
 Then import it.
 
-```js
-import VT from 'variable-type'; // ES6
-var VT = require('variable-type'); // ES5 with npm
+```typescript
+import VT from 'variable-type'; // ES6/Typescript
```
Comment (hustcc) [inline README.md]: remove the comment

// ES6/Typescript

File: jest.config.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+    roots: ['tests'],
+    preset: 'ts-jest',
+    testEnvironment: 'node',
+    collectCoverageFrom: ['src/**/*.ts'],
+    coveragePathIgnorePatterns: ['src/*.d.ts']
+};
```
Comment (hustcc) [inline jest.config.js]: write jest config in package.json

File: package-lock.json [hunk 5]
```diff
@@ -0,0 +1,5522 @@
+{
```
Comment (hustcc) [inline package-lock.json]: remove lock file from git.

File: package.json [hunk 6]
```diff
@@ -1,12 +1,14 @@
 {
   "name": "variable-type",
-  "version": "0.2.1",
+  "version": "1.0.0",
```
Comment (hustcc) [inline package.json:3]: 0.3.0 is the next break change version.

File: package.json [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
   "description": "Runtime type checking for variable and similar objects.",
-  "main": "src/index.js",
+  "main": "dist/src/index.js",
   "scripts": {
-    "coverage": "nyc npm test && nyc report --reporter=lcov",
-    "test": "mocha tests/test.js",
-    "perf": "node perf/benchmark.js"
+    "coverage": "jest --coverage",
+    "test": "jest",
+    "test:watch": "jest --coverage --watchAll",
```
Comment (hustcc) [inline package.json]: remove watch scripts

File: src/Type.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+  check: TypeChecker = (variable: any): boolean => {
+    try {
+      return this.type(variable);
+    } catch {}
+    return false;
+  };
+
+  optional: () => Type = (): Type => {
+    return _optional(this);
```
Comment (hustcc) [inline src/Type.ts:24]: why move _optional into another file.

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import instanceOf from './types/instanceOf';
+import not from './types/not';
+import or from './types/or';
+import shape from './types/shape';
+import typeOf from './types/typeOf';
+
+type FunctionThatReturnsType = (toCheck: any) => Type;
+
+export type VariableType = {
+    undefined: Type;
```
Comment (hustcc) [inline src/index.ts]: add readonly keyword.

File: src/types/apply.ts [hunk 10]
```diff
@@ -2,13 +2,13 @@
  * Created by hustcc on 17/08/01.
  */
 
-var Type = require('../Type');
+import Type, {TypeChecker} from '../Type';
```
Comment (hustcc) [inline src/types/apply.ts]: { TypeChecker } 

with space.

File: tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    "sourceMap": true,
+    "strict": true,
+    "target": "es2019"
+  },
+  "include": [
+    "src/**/*",
+    "tests/**/*",
+    "perf/**/*"
+  ]
+}
```
Comment (hustcc) [inline tsconfig.json]:
```ts
{
  "compilerOptions": {
    "outDir": "lib",
    "module": "commonjs",
    "target": "es5",
    "moduleResolution": "node",
    "experimentalDecorators": true,
    "declaration": true,
    "sourceMap": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "pretty": true,
    "lib": ["dom", "esnext"],
    "skipLibCheck": true,
    "sourceRoot": "src"
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules"]
}
```

- [ ] default build into lib
- [ ] default build commonjs es5
- [ ] ts-node can run ts file without build them into js

**Comment by hustcc** [comment]
Thanks very much for your work. I have submit some cr.

File: src/Type.ts (hunk 8 above)
Comment (ryantate13) [reply src/Type.ts:24]: That file showed 0% coverage if I didn't and it was duplicated code otherwise.

File: package.json (hunk 6 above)
Comment (ryantate13) [reply package.json:3]: Agreed, this lib is pretty stable though, don't you think? Worthy of a 1.0 IMO. I've been using it for years and it does what I expect it to which is replicate the prop-types api effectively for use server-side. I don't have a strong opinion either way honestly but I think for some team members, seeing a pre-1.0 version number can cause concern.

**Comment by ryantate13** [comment]
These notes should be resolved except the 1.0 version, see comment above

File: package.json [hunk 12]
```diff
@@ -1,12 +1,13 @@
 {
   "name": "variable-type",
-  "version": "0.2.1",
+  "version": "1.0.0",
   "description": "Runtime type checking for variable and similar objects.",
-  "main": "src/index.js",
+  "main": "dist/src/index.js",
```
Comment (hustcc) [inline package.json]: this is not good.

Comment (ryantate13) [reply package.json]: Agreed. We can switch to lib and update .npmignore but what do you think about this approach? Basically uses make and docker to build/publish and only the lib folder winds up in the package.

https://github.com/ryantate13/variable-type/tree/docker_build

Comment (hustcc) [reply package.json]: github action can do this, no need to use docker.

We just want:

1. tsc build into lib folder
2. perf / test do not build to js

**Comment by ryantate13** [comment]
Looks like that's the way it works now with the recent update to the tsconfig. I'm not familiar with how your GitHub action is set up though. Do you want main to be `lib/index.js` or just `index.js`? 🤔

**Comment by hustcc** [comment]
use `lib/index.js`, thanks.

**Comment by ryantate13** [comment]
We should be good to go now

**Comment by hustcc** [comment]
`git rm package-lock.json`

**Comment by ryantate13** [comment]
Done ✔️

**Comment by hustcc** [comment]
```diff
+ variable-type@1.0.0
```

**Pull request merged**